### PR TITLE
Add CLI based plugin discovery and installation / removal

### LINKED ISSUE #2861: DX: Improve serverless plugins discoverability & install process
# This is a Feature Proposal

## Description

Add a `serverless plugins` command so that Serverless plugins can be searched and installed (wrapper around `npm`) through the CLI. This is a big DX gain as the user doesn't need to leave the CLI to search and install plugins (it feels less nodeish).

@DavidWells and I also talked about way to offer a website where one can search through all the Serverless plugins and services. He's also working on an integration in the dashboard (so the `plugins.json` file (see below) is the single point of truth for all Serverless plugins).

**search**

The `--search` option simply looks in the registry here https://github.com/serverless/community-plugins/blob/master/plugins.json and searches through the plugins.

**install**

The `--install` option will call `npm install --save` with the name of the plugin and adds the plugin to the plugin section of the services `serverless.yml` file. A message will be shown that the user should check if the order of the plugin addition is correct (the newly installed plugin will be pushed to the end of the `plugins` array).

**remove**

The `--remove` option will call `npm uninstall --save` and removes the plugin from the services `serverless.yml` file.

---

This new `plugins` command might be reused by #2821 / #1260.

### PR DESCRIPTION
**Pull request by pmuens**
## What did you implement:

Closes #2861

This PR adds the possibility to search for plugins form the official plugin registry (https://github.com/serverless/plugins) via the CLI.

Furthermore one can easily install the plugin via the CLI. It'll also be added to the services file automatically 🎉 💃 

## How did you implement it:

Introduce a `plugin` plugin which has the commands:

- `serverless plugin list`
- `serverless plugin search --query webpack`
- `serverless plugin install --name serverless-webpack`
- `serverless plugin uninstall --name serverless-webpack`

## How can we verify it:

Checkout the branch and run one of the commands above (e.g. `serverless plugin search -q webpack`). Furthermore you can look at the corresponding tests.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

**Comment by dougmoscrop** [comment]
Thoughts on `--save-dev` (and checking devDependencies) instead of `--save`?

**Comment by pmuens** [comment]
@dougmoscrop thanks for chiming in!

That's a good suggestion (I think @nikgraf came up with something similar in #2861). I personally have always used --save when adding plugins.

What are the upsides and downsides of using `--save-dev` compared to `--save` (if any)?

**Comment by pmuens** [comment]
Forget about the comment above. It was definitely a brain lag (I wrote it in a hurry).
It's definitely reasonable to add the plugin to the `devDependencies` since all the plugins enhance the Serverless Framework and are therefore dev tools.

Thanks for the comment! I'll update the PR accordingly! 👍 💯

**Comment by pmuens** [comment]
@nikgraf @mthenw @eahefnawy would like to get this into the release tomorrow.

I updated the code so that the `etc` dir @eahefnawy reported won't show up anymore. Would be cool if you could give it a quick spin.

File: lib/plugins/plugin/plugin.js [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+            fse.writeJsonSync(packageJsonFilePath, packageJsonFileContent);
+          }
+
+          // install the package through npm
+          this.serverless.cli
+            .log(`Installing plugin "${pluginName}" (this might take a few seconds...)`);
+
+          childProcess
+            .execSync(`npm install --save-dev ${pluginName}`, {
```
Comment (nikgraf) [inline lib/plugins/plugin/plugin.js]: We shouldn't do this, but rather print:

```
Updated your package.json & serverless.yml…
To download the plugin please run `npm install` or `yarn install`
```

We also could check if a yarn.lock file exists and then run yarn …

/cc @mthenw thoughts?

File: lib/plugins/plugin/plugin.js [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+        const servicePath = this.serverless.config.servicePath;
+        const packageJsonFilePath = path.join(servicePath, 'package.json');
+
+        // uninstall the package through npm
+        this.serverless.cli
+          .log(`Uninstalling plugin "${pluginName}" (this might take a few seconds...)`);
+
+        childProcess
+          .execSync(`npm uninstall --save-dev ${pluginName}`, {
```
Comment (nikgraf) [inline lib/plugins/plugin/plugin.js]: Same as above …

File: lib/plugins/plugin/plugin.js (hunk 1 above)
Comment (nikgraf) [inline lib/plugins/plugin/plugin.js]: Ok, just realised we somehow need to find the latest stable version. We could use the npm api directly without going through the CLI. I might not have `npm` installed. We then can add it to the package.json.

Comment (DavidWells) [inline lib/plugins/plugin/plugin.js]: Everyone should have NPM installed. 

We can do something like:
1. Try/catch run npm install
2. If fails try/catch yarn add (maybe do a check for yarn like nik mentions before step 1 and use it instead)
3. If fails good old fashioned download plugin tarball

**Comment by dougmoscrop** [comment]
That's not the same thing as what Nik is getting at..

**Comment by eahefnawy** [comment]
Will be moved to an external plugin (not in core)

**Pull request closed without merging**